### PR TITLE
dialects: (arith) Preserve special values when folding division by zero

### PR DESCRIPTION
**Pull request by AdemVessell**
The zero-divisor path in `_fold_const_operation` determines infinity's sign from the numerator alone and turns a NaN numerator into infinity. A negative-zero denominator can therefore reverse a subsequent comparison.

For example, canonicalizing this no-fastmath function produces a positive-infinity quotient and returns true before this fix; the quotient should be negative infinity and the comparison false:

```mlir
builtin.module {
  func.func @division_branch() -> i1 {
    %one = arith.constant 1.0 : f64
    %negative_zero = arith.constant -0.0 : f64
    %quotient = arith.divf %one, %negative_zero : f64
    %greater = arith.cmpf ogt, %quotient, %one : f64
    func.return %greater : i1
  }
}
```

For a zero divisor, the patch returns NaN when the numerator is zero or NaN, and otherwise determines infinity's sign from both operands. The nonzero-divisor path is unchanged.

Following review, all 22 cases are in `tests/filecheck/mlir-conversion/with-mlir/dialects/arith/divf_constant_folding.mlir`, with two RUN directives: one for xDSL canonicalization and one for MLIR canonicalization. The final xDSL invocation on the MLIR path only normalizes printing and runs no passes. The separate relative-path wrapper has been removed.

The f64 cases cover finite/infinite numerator signs, both zero-divisor signs, NaN, zero/zero and ordinary division controls. The f16/f32 cases cover negative-zero divisors and NaN propagation. This consolidated integration test uses the existing with-mlir test gate.

These checks establish canonical NaN results, not preservation of arbitrary NaN payloads or signs. The arithmetic repair is unchanged by this test revision.

Earlier validation before the September 12 test-only consolidation (ARM64 macOS):

- Shared expectations passed with xDSL and native MLIR 20.1.8 and 21.1.8.
- The same fixture fails with the original folder on `1.0 / -0.0`, which produces positive instead of negative infinity.
- 1,130 focused pytest cases pass on each of Python 3.12.13 and 3.14.4; four existing FileCheck specifications and the new xDSL fixture pass on both.
- Ruff, formatting and targeted Pyright pass.
- The separate 81-case native ARM64 division comparison still has zero value/comparison mismatches. This is supplementary host-specific evidence, not a portable C proof or an additional shipped test harness.

September 12 follow-up: all 22 function bodies and CHECK assertions are unchanged. The consolidated test passes through lit with MLIR 21.1.8 and 22.1.2 on Python 3.12; its commands also pass directly on Python 3.14. Restoring the old folder in a disposable interpreter still makes the fixture fail. The production repair is unchanged. The earlier pytest totals above were not rerun for this file move.

```sh
python -m pytest -q tests/dialects/test_arith.py tests/interpreters/test_arith_interpreter.py tests/transforms/test_canonicalization.py tests/test_folder.py tests/test_ir.py tests/pattern_rewriter/test_pattern_rewriter.py
XDSL_MLIR_OPT=/path/to/mlir-opt lit -v tests/filecheck/mlir-conversion/with-mlir/dialects/arith/divf_constant_folding.mlir
```

Local validation is not the full upstream CI matrix. Developed with AI assistance; the reproducer, patch and listed checks were executed locally.

### CODE REVIEW

File: tests/transforms/test_arith_float_folding.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        (6.0, 2.0, 3.0),
+        (6.0, -2.0, -3.0),
+    ],
+)
+def test_divf_constant_folding_special_values(
+    float_type: builtin.Float16Type | builtin.Float32Type | builtin.Float64Type,
+    left: float,
+    right: float,
+    expected: float,
+):
```
Comment (alexarice) [inline tests/transforms/test_arith_float_folding.py]: To me, this pytest feels slightly over engineered and depends on a fair amount of infrastructure. In general for more integration like tests, we usually use `filecheck`, which I feel would fit well for this sort of test. Doing this would also allow us to check that we match the behaviour of MLIR exactly.

**Review by alexarice — COMMENTED** [review]
The logic looks good, though I'd love to see an MLIR integration test for this, as I'm aware there's lots of corner cases with float logic.

**Comment by AdemVessell** [comment]
Thanks, Alex — agreed. Replaced the interpreter-based pytest in 79eea1064ba5814137806cacecbe4306e2530052 with a FileCheck fixture and a small MLIR integration wrapper.

Both canonicalizers now use the same expected results. The main fixture covers all sign combinations for finite/infinite numerators divided by zero, NaN/zero, zero/zero and ordinary division controls in f64, plus negative-zero and NaN regressions in f16/f32. The wrapper runs MLIR canonicalization; its final xDSL invocation is printing-only, with no optimization passes. Normal xDSL testing still works without MLIR.

All 22 cases pass locally with xDSL and native MLIR 20.1.8 and 21.1.8. The repository's 22.1.2 CI check is still pending for this update. Restoring the original folder makes the fixture fail on `1.0 / -0.0` (positive instead of negative infinity). The 1,130 focused pytest cases and existing folding FileChecks also pass on Python 3.12.13 and 3.14.4.

The production repair is unchanged. The NaN checks are for canonical NaN results, not arbitrary payload/sign preservation. I've updated the PR description with the new test paths and commands.

**Review by alexarice — COMMENTED** [review]
Sorry, I should have specified, by MLIR integration test I mean a filecheck test in `tests/filecheck/mlir-integration/with-mlir`. You can then simply have two `RUN` directives at the top of the file, one for `xdsl-opt` and one for `$XDSL_MLIR_OPT`

**Comment by alexarice** [comment]
Ah, I actually see now that your current test is in the right place, though I think it's better to simply move all the testing here and to not use the current relative path. Sorry for the confusion

**Comment by AdemVessell** [comment]
Thanks, that makes sense. Consolidated all 22 cases in the existing with-mlir test in 7f7e983 and removed the relative-path wrapper. It now has two RUN directives, with the final xdsl-opt on the MLIR path used only to normalize printing. Both runs pass with MLIR 21.1.8 and 22.1.2. The cases, CHECK assertions and production fix are unchanged.

**Review by alexarice — APPROVED** [review]
This looks great to me, thanks for iterating
